### PR TITLE
Database test

### LINKED ISSUE #227: Create mock database and write test cases for CRUD classes
1. Read our [Contributor Guidelines](https://github.com/CarmineOptions/derisk-research/blob/master/CONTRIBUTING.md) and [README](https://github.com/CarmineOptions/derisk-research/blob/master/README.md).
2. In this [tests](https://github.com/CarmineOptions/derisk-research/tree/master/data_handler/tests) folder, create `confest.py` file where you will mock db connection using `pytest` package
3. Create new folder `db_test` in [tests](https://github.com/CarmineOptions/derisk-research/tree/master/data_handler/tests) folder and add there next files `test_db_connector.py` and `test_initializer_db_connector.py` in these files, create test cases for next classes: 
- [DBConnector](https://github.com/CarmineOptions/derisk-research/blob/master/data_handler/db/crud.py#L25)
- [InitializerDBConnector](https://github.com/CarmineOptions/derisk-research/blob/master/data_handler/db/crud.py#L25)
4. Please, create negative and positive scenarios. So it should test not only success execution but also raise error to proper handling

### PR DESCRIPTION
**Pull request by NueloSE**
Fixes #227 

mock database and InitializerDBConnector are ready for review still workin on DBConnector

### CODE REVIEW

**Comment by djeck1432** [comment]
@NueloSE Okay, thank you for update

File: data_handler/tests/conftest.py [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    
+    session = SessionLocal()
+    
+    yield session
+    
+    session.close()
+
+@pytest.fixture(scope="function")
+def mock_db_connector(mock_session):
+    return DBConnector(db_url="sqlite:///:memory:")
```
Comment (djeck1432) [inline data_handler/tests/conftest.py:26]: Please, don't use any db, just simple mock database connection,  example of how to do it
https://github.com/djeck1432/spotnet/blob/main/web_app/tests/conftest.py

File: data_handler/tests/db_test/test_db_connector.py [hunk 2]
```diff
@@ -0,0 +1,311 @@
...
+from uuid import uuid4
+from db.crud import DBConnector
+from db.models import LoanState, InterestRate, OrderBookModel, ZkLendCollateralDebt, HashtackCollateralDebt
+
+class TestDBConnector:
+    @pytest.fixture(autouse=True)
+    def db_connector(self):
+        return DBConnector(db_url="sqlite:///:memory:")
+
+    def test_write_to_db(self, db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_db_connector.py:13]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_db_connector.py [hunk 3]
```diff
@@ -0,0 +1,311 @@
+# tests/db_test/test_db_connector.py
+
+import pytest
+from uuid import uuid4
+from db.crud import DBConnector
+from db.models import LoanState, InterestRate, OrderBookModel, ZkLendCollateralDebt, HashtackCollateralDebt
+
+class TestDBConnector:
+    @pytest.fixture(autouse=True)
+    def db_connector(self):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_db_connector.py:10]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_db_connector.py [hunk 4]
```diff
@@ -0,0 +1,311 @@
...
+                collateral={},
+                debt={},
+                timestamp=None,
+                block=None,
+                deposit={}
+            )
+            db_connector.write_to_db(incomplete_loan_state)
+
+    def test_get_object(self, db_connector):
+        loan_state = LoanState(
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_db_connector.py:43]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_db_connector.py [hunk 5]
```diff
@@ -0,0 +1,311 @@
...
+        db_connector.write_to_db(loan_state)
+        
+        retrieved_loan_state = db_connector.get_object(model=LoanState, obj_id=loan_state.id)
+        assert retrieved_loan_state == loan_state
+        # Negative scenario
+        non_existent_id = uuid4()
+        result = db_connector.get_object(model=LoanState, obj_id=non_existent_id)
+        assert result is None
+
+    def test_delete_object(self, db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_db_connector.py:62]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_db_connector.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+        
+        retrieved_loan_state = db_connector.get_object(model=LoanState, obj_id=loan_state.id)
+        assert retrieved_loan_state is None
+
+        # Negative scenario: Attempt to delete a non-existent object
+        with pytest.raises(Exception):
+            db_connector.delete_object(model=LoanState, obj_id=uuid4())
+
+
+    def test_get_latest_block(self, db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_db_connector.py:85]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_db_connector.py [hunk 7]
```diff
@@ -0,0 +1,311 @@
...
+        
+        latest_block = db_connector.get_last_block(protocol_id="test_protocol")
+        assert latest_block == 67890
+
+        # Negative scenario: No blocks for given protocol should return None
+        non_existent_protocol = "non_existent_protocol"
+        latest_block = db_connector.get_last_block(protocol_id=non_existent_protocol)
+        assert latest_block is None
+
+    def test_write_batch_to_db(self, db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_db_connector.py:118]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_db_connector.py [hunk 8]
```diff
@@ -0,0 +1,311 @@
...
+        assert any(ls.user == "user1" for ls in retrieved_loan_states)
+        assert any(ls.user == "user2" for ls in retrieved_loan_states)
+
+        # Negative scenario: Attempt to write an invalid batch with an incomplete LoanState
+        invalid_loan_states = loan_states + [LoanState(id=uuid4())]  # Incomplete LoanState
+        with pytest.raises(Exception):
+            db_connector.write_batch_to_db(invalid_loan_states)
+
+
+    def test_get_latest_order_book(self, db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_db_connector.py:154]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_db_connector.py [hunk 9]
```diff
@@ -0,0 +1,311 @@
...
+        db_connector.write_to_db(order_book)
+        
+        retrieved_order_book = db_connector.get_latest_order_book(dex="Uniswap", token_a="ETH", token_b="USDC")
+        assert retrieved_order_book == order_book
+
+        # Negative scenario: Retrieve order book with non-existent token pair
+        result = db_connector.get_latest_order_book(dex="Uniswap", token_a="BTC", token_b="ETH")
+        assert result is None
+
+    def test_get_unique_users_last_block_objects(self, db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_db_connector.py:175]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_db_connector.py [hunk 10]
```diff
@@ -0,0 +1,311 @@
...
+        )
+        db_connector.write_to_db(loan_state1)
+        db_connector.write_to_db(loan_state2)
+        
+        unique_users_last_blocks = db_connector.get_unique_users_last_block_objects(protocol_id="test_protocol")
+        assert len(unique_users_last_blocks) == 2
+        assert any(ls.user == "user1" for ls in unique_users_last_blocks)
+        assert any(ls.user == "user2" for ls in unique_users_last_blocks)
+
+    def test_get_last_interest_rate_record_by_protocol_id(self, db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_db_connector.py:204]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_db_connector.py [hunk 11]
```diff
@@ -0,0 +1,311 @@
...
+            debt={"token": "75"},
+            timestamp=1643723400,
+            block=67890
+        )
+        db_connector.write_to_db(interest_rate)
+        
+        last_interest_rate = db_connector.get_last_interest_rate_record_by_protocol_id("test_protocol")
+        assert last_interest_rate == interest_rate
+
+    def test_get_interest_rate_by_block(self, db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_db_connector.py:218]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_db_connector.py [hunk 12]
```diff
@@ -0,0 +1,311 @@
...
+        
+        retrieved_interest_rate = db_connector.get_interest_rate_by_block(block_number=67890, protocol_id="test_protocol")
+        assert retrieved_interest_rate == interest_rate
+
+         # Negative scenario: Attempt to retrieve interest rate for non-existent block
+        non_existent_block = 99999
+        result = db_connector.get_interest_rate_by_block(block_number=non_existent_block, protocol_id="test_protocol")
+        assert result is None
+
+    def test_get_all_block_records(self, db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_db_connector.py:237]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_db_connector.py [hunk 13]
```diff
@@ -0,0 +1,311 @@
...
+        ]
+        db_connector.write_batch_to_db(loan_states)
+        
+        all_block_records = db_connector.get_all_block_records(model=LoanState)
+        assert len(all_block_records) == 2
+        assert all_block_records[0].collateral == {"token": "100"}
+        assert all_block_records[1].collateral == {"token": "150"}
+
+
+def test_write_loan_states_to_db(self, db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_db_connector.py:268]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_initializer_db_connector.py [hunk 14]
```diff
@@ -0,0 +1,80 @@
+# tests/db_test/test_initializer_db_connector.py
+
+import pytest
+from uuid import uuid4
+from db.crud import InitializerDBConnector
+from db.models import ZkLendCollateralDebt, HashtackCollateralDebt
+
+class TestInitializerDBConnector:
+    @pytest.fixture(autouse=True)
+    def initializer_db_connector(self):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_initializer_db_connector.py:10]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_initializer_db_connector.py [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+        return InitializerDBConnector(db_url="sqlite:///:memory:")
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_initializer_db_connector.py:11]: don't use sqllite, just mock it

File: data_handler/tests/db_test/test_initializer_db_connector.py [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+
+    def test_get_zklend_by_user_ids(self, initializer_db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_initializer_db_connector.py:13]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_initializer_db_connector.py [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+            collateral={"token": 200},
+            debt={"token": 75}
+        )
+
+        retrieved_records = initializer_db_connector.get_zklend_by_user_ids([user_id1, user_id2])
+        assert len(retrieved_records) == 2
+        assert any(record.user_id == user_id1 for record in retrieved_records)
+        assert any(record.user_id == user_id2 for record in retrieved_records)
+
+    def test_save_debt_category(self, initializer_db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_initializer_db_connector.py:36]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_initializer_db_connector.py [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+        retrieved_record = initializer_db_connector.get_hashtack_by_loan_ids([loan_id], version=1)[0]
+        assert retrieved_record.loan_id == loan_id
+        assert retrieved_record.debt_category == "test_category"
+        assert retrieved_record.collateral == {"token": 150}
+        assert retrieved_record.debt == {"token": 75}
+        assert retrieved_record.original_collateral == {"token": 100}
+        assert retrieved_record.borrowed_collateral == {"token": 50}
+
+
+    def test_get_zklend_by_user_ids_empty_result(self, initializer_db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_initializer_db_connector.py:60]: add docstring and annotation type and return type

File: data_handler/tests/db_test/test_initializer_db_connector.py [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+        retrieved_records = initializer_db_connector.get_zklend_by_user_ids(["nonexistent_user"])
+        assert len(retrieved_records) == 0
+
+    def test_save_debt_category_invalid_version(self, initializer_db_connector):
```
Comment (djeck1432) [inline data_handler/tests/db_test/test_initializer_db_connector.py:64]: add docstring and annotation type and return type

**Review by djeck1432 — CHANGES_REQUESTED** [review]

**Comment by djeck1432** [comment]
@NueloSE Any update?

**Comment by djeck1432** [comment]
@NueloSE Please, fix all comments till tomorrow morning, thank you

**Comment by djeck1432** [comment]
@NueloSE Please, finish this task as soon as possible, this task blocking us to publish new issues, thank you for your understanding

**Comment by NueloSE** [comment]
@djeck1432 am on it implement requested changes before end of today

**Comment by djeck1432** [comment]
@NueloSE Just checked your github, you are contributing to other project, others devs waiting when you finish this task. this task should be done today, otherwise I will be merge it and you will have just half of reward. 
Please, don't do that next time.

**Pull request closed without merging**